bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java

```diff
@@ -0,0 +1,491 @@
+                    tivoConfigData.getCfgIdentifier(), iL, tivoConfigData.getCfgNumConnRetry());
+
+            // Sort out the socket connection
+            if (connSocketConnect()) {
+                logger.debug(" connTivoConnect '{}' - Socket created / connection made.",
+                        tivoConfigData.getCfgIdentifier());
+                StreamReader streamReader = this.streamReader;
+                if (streamReader != null && streamReader.isAlive()) {
+                    return true;
+                }
+            } else {
+                logger.debug(" connTivoConnect '{}' - Socket creation failed.", tivoConfigData.getCfgIdentifier());
+                TiVoHandler tivoHandler = this.tivoHandler;
+                if (tivoHandler != null) {
+                    tivoHandler.setStatusOffline();
+                }
+            }
+            // Sleep and retry
+            doNappTime();
+        }
+        return false;
+    }
+
+    /**
+     * {@link connTivoDisconnect} conditionally closes the Socket connection. When 'keep connection open' or 'channel
+     * scanning' is true, the disconnection process is ignored. Disconnect can be forced by setting forceDisconnect to
+     * true.
+     *
+     * @param forceDisconnect true = forces a disconnection , false = disconnects in specific situations
+     */
+    public void connTivoDisconnect(boolean forceDisconnect) {
+        if (forceDisconnect) {
+            connSocketDisconnect();
+        } else {
+            if (!tivoConfigData.isCfgKeepConnOpen()) {
+                doNappTime();
+                connSocketDisconnect();
+            }
+        }
+    }
+
+    /**
+     * {@link connTivoReconnect} disconnect and reconnect the socket connection to the TiVo.
+     *
+     * @return boolean true = connection succeeded, false = connection failed
+     */
+    public boolean connTivoReconnect() {
+        connSocketDisconnect();
+        doNappTime();
+        return connTivoConnect();
+    }
+
+    /**
+     * {@link connSocketDisconnect} cleanly closes the socket connection and dependent objects
+     *
+     */
+    private void connSocketDisconnect() {
+        TiVoHandler tivoHandler = this.tivoHandler;
+        StreamReader streamReader = this.streamReader;
+        PrintStream streamWriter = this.streamWriter;
+        Socket tivoSocket = this.tivoSocket;
+
+        logger.debug(" connTivoSocket '{}' - requested to disconnect/cleanup connection objects",
+                tivoConfigData.getCfgIdentifier());
+
+        // if isCfgKeepConnOpen = false, don't set status to OFFLINE since the socket is closed after each command
+        if (tivoHandler != null && tivoConfigData.isCfgKeepConnOpen()) {
+            tivoHandler.setStatusOffline();
+        }
+
+        try {
+            if (streamReader != null) {
+                while (streamReader.isAlive()) {
+                    streamReader.stopReader();
+                }
```

Comment:
Stopping of the Thread seems quite cumbersome. You could use `Thread.join(timeout)` here. That would also make the CountDownLatch unnecessary.

Reply:
Fixed by removing the stopReader() method an moving the logic up to connSocketDisconnect()